Applications/Help/ManualModel.cpp

```diff
@@ -133,10 +136,42 @@ int ManualModel::column_count(const GUI::ModelIndex&) const
     return 1;
 }
 
+static bool occurs_in_file(const String& term, const String& path)
+{
+    auto file_or_error = Core::File::open(path, Core::IODevice::ReadOnly);
+    if (file_or_error.is_error())
+        return false;
+
+    auto file = file_or_error.value();
+
+    for (;;) {
+        auto buffer = file->read_line(4096);
+        if (buffer.is_null())
+            return false;
+        if (String { (const char*)buffer.data() }.contains(term))
```

Comment:
This is wrong for two reasons:
* You can't assume that the file doesn't have 0 bytes, nor that `read_line()` will return a 0-terminated buffer (in fact one of the paths there does 0-terminate the buffer, and the other doesn't, and that is wrong too...)
* If the line is long and doesn't fit into the provided buffer size (interesting to see how robust `Core::IODevice::read_line()` is for that case...), you'd miss the occurrence at the edge of the reads. Example: searching for "needle", first read returns "... ne", second read returns "edle ..."

Reply:
> You can't assume that the file doesn't have 0 bytes in it, nor that read_line() will return a 0-terminated buffer

Perfect, an excuse to add `memmem` to `AK::ByteBuffer`.
that class is my favourite buffer type :P 

> If the line is long and doesn't fit into the provided buffer size

That's fair. I'd really rather just map the file and do a single pass over it, but I'm not very familiar with `MappedFile` and whatever interfaces it provides for this use, if any.

Reply:
Yeah, using mmap for this sounds much better. `MappedFile` gives you data + size, from which you can construct a string view, and go from there. Bonus point: no copying, other than the initial paging in from the inode.